OrderingSuite.testBinarySearchOnDict fails sporadically

crank up the number of tests in Prop.scala to replicate, or use this one:

```
> ARG_0: (dict<call, float64>,Map(3 -> 1.4756407611933507E308, 0 -> 46.25904804767313, 1 -> 54.68552327526001),12)

java.lang.AssertionError: java.lang.ArrayIndexOutOfBoundsException: 3

	at is.hail.check.GenProp1$$anonfun$apply$1.apply$mcVI$sp(Prop.scala:38)
	at scala.collection.immutable.Range.foreach$mVc$sp(Range.scala:160)
	at is.hail.check.GenProp1.apply(Prop.scala:26)
	at is.hail.check.Prop.check(Prop.scala:19)
	at is.hail.expr.ir.OrderingSuite.testBinarySearchOnDict(OrderingSuite.scala:357)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:696)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:882)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1189)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:124)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at org.testng.TestRunner.privateRun(TestRunner.java:767)
	at org.testng.TestRunner.run(TestRunner.java:617)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:348)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:343)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:305)
	at org.testng.SuiteRunner.run(SuiteRunner.java:254)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1224)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1149)
	at org.testng.TestNG.run(TestNG.java:1057)
	at org.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:72)
	at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:123)
Caused by: java.lang.ArrayIndexOutOfBoundsException: 3
	at scala.collection.mutable.WrappedArray$ofRef.apply(WrappedArray.scala:127)
	at is.hail.expr.ir.OrderingSuite$$anonfun$33$$anonfun$apply$19.getKey$1(OrderingSuite.scala:345)
	at is.hail.expr.ir.OrderingSuite$$anonfun$33$$anonfun$apply$19.apply(OrderingSuite.scala:347)
	at is.hail.expr.ir.OrderingSuite$$anonfun$33$$anonfun$apply$19.apply(OrderingSuite.scala:318)
	at is.hail.utils.package$.using(package.scala:613)
	at is.hail.annotations.Region$.scoped(Region.scala:13)
	at is.hail.expr.ir.OrderingSuite$$anonfun$33.apply(OrderingSuite.scala:318)
	at is.hail.expr.ir.OrderingSuite$$anonfun$33.apply(OrderingSuite.scala:314)
	at is.hail.check.GenProp1$$anonfun$apply$1$$anonfun$1.apply$mcZ$sp(Prop.scala:28)
	at is.hail.check.GenProp1$$anonfun$apply$1$$anonfun$1.apply(Prop.scala:28)
	at is.hail.check.GenProp1$$anonfun$apply$1$$anonfun$1.apply(Prop.scala:28)
	at scala.util.Try$.apply(Try.scala:192)
	at is.hail.check.GenProp1$$anonfun$apply$1.apply$mcVI$sp(Prop.scala:28)
	... 27 more
```

--- a/hail/src/test/scala/is/hail/expr/ir/OrderingSuite.scala
+++ b/hail/src/test/scala/is/hail/expr/ir/OrderingSuite.scala
@@ -342,16 +342,20 @@ class OrderingSuite extends SparkSuite {
         val f = fb.resultWithIndex()(0)
         val closestI = f(region, soff, eoff)
 
-        def getKey(i: Int) = asArray(i).asInstanceOf[Row].get(0)
+        if (closestI == asArray.length) {
+          !dict.contains(key) ==> asArray.forall { keyI =>
+            val otherKey = keyI.asInstanceOf[Row].get(0)
+            pDict.keyType.virtualType.ordering.compare(key, otherKey) > 0
+          }
+        } else {
+          def getKey(i: Int) = asArray(i).asInstanceOf[Row].get(0)
+          val maybeEqual = getKey(closestI)
+          val closestIIsClosest =
+            (pDict.keyType.virtualType.ordering.compare(key, maybeEqual) <= 0 || closestI == dict.size - 1) &&
+              (closestI == 0 || pDict.keyType.virtualType.ordering.compare(key, getKey(closestI - 1)) > 0)
 
-        val maybeEqual = getKey(closestI)
-
-        val closestIIsClosest =
-          (pDict.keyType.virtualType.ordering.compare(key, maybeEqual) <= 0 || closestI == dict.size - 1) &&
-            (closestI == 0 || pDict.keyType.virtualType.ordering.compare(key, getKey(closestI - 1)) > 0)
-
-        dict.contains(key) ==> (key == maybeEqual) && closestIIsClosest
-
+          dict.contains(key) ==> (key == maybeEqual) && closestIIsClosest
+        }
       }
     }
     p.check()

